packages/ui/src/components/Node/ItemCollection.ts

```diff
@@ -1,26 +1,83 @@
 import { ItemValue } from '@data-story/core/*'
 
+type JSONValue = string | number | boolean | {[key: string]: JSONValue} | JSONValue[];
+
 export class ItemCollection {
-  constructor(public items: ItemValue[]) {}
-
-  toTable() {
-    let headers: string[] = []
-    let rows: ItemValue[] = []
-
-    // Extract headers and remove duplicates
-    headers = Array.from(new Set(this.items
-      .map((item) => Object.keys(item))
-      .flat()));
-
-    // Extract rows
-    rows = this.items.map((item) => {
-      const row: ItemValue[] = []
-      headers.forEach((header) => {
-        row.push(item[header])
-      })
-      return row
-    })
-
-    return { headers, rows }
+  constructor(public items: ItemValue[]) {
+  }
+
+  private serializeArrayContent(arrayContent: unknown[]): string {
+    return JSON.stringify(arrayContent).replace(/\n/g, '');
   }
-}
\ No newline at end of file
+
+  public toTable() {
+    const headers: Set<string> = new Set();
+    const rows: (string | undefined)[][] = [];
+
+    /**
+     * @description recursively build headers
+     */
+    const buildHeaders = (entry: {[key: string]: JSONValue}, prefix: string = '') => {
+      Object.entries(entry).forEach(([key, value]) => {
+        const newKey = prefix ? `${prefix}.${key}` : key;
+        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
+          buildHeaders(value as {[key: string]: JSONValue}, newKey);
+        } else {
+          headers.add(newKey);
+        }
+      });
+    };
+
+    /**
+     * @description recursive data to build headers
+     */
+    this.items.forEach(item => {
+      if (typeof item === 'object' && item !== null && !Array.isArray(item)) {
+        buildHeaders(item);
+      }
+    });
+
+    /**
+     * @description get value by header's path
+     */
+    const getValueByPath = (object: ItemValue, path: string[]): string | undefined => {
```

Comment:
Not sure if usable here, but can be good to know this utility exists: packages/core/src/utils/get.ts

Reply:
`getValueByPath` includes some special functionalities for handling types of array and object values

Reply:
👍